src/pip/_internal/operations/install/wheel.py

```diff
@@ -420,6 +421,15 @@ def __init__(self, src_record_path, dest_path, zip_file):
         self._zip_file = zip_file
         self.changed = False
 
+    def _getinfo(self):
+        # type: () -> ZipInfo
+        if not PY2:
+            return self._zip_file.getinfo(self.src_record_path)
```

Comment:
Do you know why this work on python 3? Is the encoding guaranteed to be utf-8 or is there a risk it depends on some environmental condition?

Reply:
Python 3 uses a flag in the ZIP header:

```
Bit 11: Language encoding flag (EFS).  If this bit is set,
        the filename and comment fields for this file
        MUST be encoded using UTF-8. (see APPENDIX D)
```

(Source: https://pkware.cachefly.net/webdocs/APPNOTE/APPNOTE-6.3.8.TXT)

Internally `ZipFile.getinfo()` (which is also used by `ZipFile.open()`) is implemented by building a dict from the ZIP’s metadata, with the keys being the file names. Python 3 would check the flag and decode the path with UTF-8 if it’s set.

Python 2, on the other hand, does not make use of the flag (I suspect the implementation actually predates the flag), and the dict is built using Python 2 `str`. I actually implemented logic to guess the encoding in d4995cb89eed0a2d348e220c6ef061b3d816e0f4, but dropped that since most of it is unnecessary if the wheel is guarenteed to use UTF-8. I can revert it if needed.

Reply:
Oh, nice. Thanks for the detailed explanation.